### PR TITLE
Feature/generated maps in pools

### LINKED ISSUE #118: Support new map assignments endpoint and map weights

### PR DESCRIPTION
**Pull request by Sheikah45**
Fixes #118

### CODE REVIEW

**Comment by Sheikah45** [comment]
Needs api with https://github.com/FAForever/faf-java-api/pull/447

**Comment by Sheikah45** [comment]
Not sure why the build is failing, it works locally

**Comment by Sheikah45** [comment]
@Brutus5000 do you have any insights should Value not be used?

**Comment by Brutus5000** [comment]
I have no clue. The error is weird. Maybe Travis changed something with the Java version.
Maybe try to upgrade Gradle?

File: src/main/java/com/faforever/moderatorclient/api/domain/MapService.java [hunk 1]
```diff
@@ -107,28 +120,50 @@
...
         for (MatchmakerQueueMapPool bracket : brackets) {
             log.info("{}", bracket);
         }
         return brackets;
     }
 
-    public void patchMapPool(MapPoolFX mapPoolFX) {
-        patchMapPool(mapPoolMapper.mapToDto(mapPoolFX));
+    public List<MapPoolAssignment> getListOfMapsInBrackets(List<MatchmakerQueueMapPool> brackets) {
+        List<String> poolIDs = brackets.stream().map(bracket -> bracket.getMapPool().getId()).collect(Collectors.toList());
```
Comment (Brutus5000) [inline src/main/java/com/faforever/moderatorclient/api/domain/MapService.java]: lienbreak before map and before collect

File: src/main/java/com/faforever/moderatorclient/api/domain/MapService.java [hunk 2]
```diff
@@ -107,28 +120,50 @@
...
+        ElideNavigatorOnCollection<MapPoolAssignment> routeBuilder = ElideNavigator.of(MapPoolAssignment.class)
+                .collection()
+                .addFilter(ElideNavigator.qBuilder().string("mapPool.id").in(poolIDs))
+                .addIncludeOnCollection("mapVersion")
+                .addIncludeOnCollection("mapVersion.map")
+                .addSortingRule("mapVersion.width", false)
+                .addSortingRule("mapVersion.map.displayName", false);
+        List<MapPoolAssignment> mapAssignments = fafApi.getAll(MapPoolAssignment.class, routeBuilder);
+        for (MapPoolAssignment poolAssignment : mapAssignments) {
+            log.info("{}", poolAssignment);
```
Comment (Brutus5000) [inline src/main/java/com/faforever/moderatorclient/api/domain/MapService.java]: Give a text, reconsider log level (debug??)

File: src/main/java/com/faforever/moderatorclient/api/domain/MapService.java [hunk 3]
```diff
@@ -177,4 +212,25 @@ public boolean doesMapVersionExist(int id) {
...
+        ResponseEntity<List<GithubGeneratorRelease>> response = restTemplate.exchange(generatorVersionsURL, HttpMethod.GET, entity, new ParameterizedTypeReference<>() {
+        });
+        List<GithubGeneratorRelease> releases = response.getBody();
+        for (GithubGeneratorRelease release : releases) {
+            ComparableVersion version = new ComparableVersion(release.getTagName());
+            if (version.compareTo(minVersion) >= 0) {
+                versions.add(version);
+            }
+        }
+        return versions;
```
Comment (Brutus5000) [inline src/main/java/com/faforever/moderatorclient/api/domain/MapService.java]: Stream with filter seems more appropriate.

File: src/main/java/com/faforever/moderatorclient/common/MapParams.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+@JsonTypeInfo(use = Id.NAME, property = "type")
+@JsonSubTypes(
+    @JsonSubTypes.Type(value = NeroxisGeneratorParams.class, name = "neroxis")
+)
+@Getter
+@Setter
+@NoArgsConstructor
+public abstract class MapParams {
+}
```
Comment (Brutus5000) [inline src/main/java/com/faforever/moderatorclient/common/MapParams.java]: Needs to be removed as soon as faf-java-commons PR is merged

File: src/main/java/com/faforever/moderatorclient/common/MapPoolAssignment.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+@Setter
+@Type("mapPoolAssignment")
+public class MapPoolAssignment extends AbstractEntity {
+  @Relationship("mapPool")
+  private MapPool mapPool;
+  @Relationship("mapVersion")
+  private MapVersion mapVersion;
+  private int weight;
+  private MapParams mapParams;
+}
```
Comment (Brutus5000) [inline src/main/java/com/faforever/moderatorclient/common/MapPoolAssignment.java]: Needs to be removed as soon as faf-java-commons PR is merged

File: src/main/java/com/faforever/moderatorclient/common/NeroxisGeneratorParams.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+@EqualsAndHashCode(callSuper = false)
+@Getter
+@Setter
+@JsonTypeName("neroxis")
+public class NeroxisGeneratorParams extends MapParams {
+  private int spawns;
+  private int size;
+  private String version;
+}
```
Comment (Brutus5000) [inline src/main/java/com/faforever/moderatorclient/common/NeroxisGeneratorParams.java]: Needs to be removed as soon as faf-java-commons PR is merged

File: src/main/java/com/faforever/moderatorclient/ui/main_window/LadderMapPoolController.java [hunk 7]
```diff
@@ -57,8 +67,16 @@
...
     public VBox addButtonsContainer;
+    public ComboBox<ComparableVersion> neroxisVersionComboBox;
+    public Spinner<String> neroxisSizeSpinner;
+    public Spinner<Integer> neroxisSpawnsSpinner;
 
-    private ObjectProperty<MapVersionFX> selectedMap = new SimpleObjectProperty<>();
+    private final ObjectProperty<MapPoolAssignmentFX> selectedMap = new SimpleObjectProperty<>();
+    private final BiPredicate<MapPoolAssignmentFX, MapPoolAssignmentFX> matchingPoolAssignmentPredicate = (assignmentFX1, assignmentFX2) ->
+            (assignmentFX1.getMapVersion() != null && assignmentFX1.getMapVersion().equals(assignmentFX2.getMapVersion()))
+                    || (assignmentFX1.getMapParams() != null && assignmentFX1.getMapParams().equals(assignmentFX2.getMapParams()));
```
Comment (Brutus5000) [inline src/main/java/com/faforever/moderatorclient/ui/main_window/LadderMapPoolController.java]: Use Objects.equals instead. Null safe and more readable

File: src/main/java/com/faforever/moderatorclient/ui/main_window/LadderMapPoolController.java [hunk 8]
```diff
@@ -133,37 +168,61 @@ private void loadMatchMakerQueue(MatchmakerQueue matchmakerQueue) {
...
+            List<MapPoolAssignmentFX> newMapPoolAssignments = bracketMapPoolAssignments.stream()
+                    .filter(mapPoolAssignment -> !oldMapPoolAssignmentsFX.contains(mapPoolAssignment))
+                    .collect(Collectors.toList());
+            List<MapPoolAssignmentFX> removedMapPoolAssignments = oldMapPoolAssignmentsFX.stream()
+                    .filter(mapPoolAssignment -> !bracketMapPoolAssignments.contains(mapPoolAssignment))
+                    .collect(Collectors.toList());
+            List<MapPoolAssignmentFX> changedMapPoolAssignments = bracketMapPoolAssignments.stream()
+                    .filter(mapPoolAssignment -> oldMapPoolAssignmentsFX.stream()
+                            .anyMatch(assignmentFX -> Objects.equals(mapPoolAssignment, assignmentFX)
+                            && !mapPoolAssignment.getWeight().equals(assignmentFX.getWeight())))
```
Comment (Brutus5000) [inline src/main/java/com/faforever/moderatorclient/ui/main_window/LadderMapPoolController.java]: Objects.equals here too

File: src/main/java/com/faforever/moderatorclient/ui/main_window/LadderMapPoolController.java [hunk 9]
```diff
@@ -176,40 +235,50 @@ private void bindListViewSelectionToSelectedMapProperty(ListView<MapVersionFX> l
...
 
     private String getBracketRatingString(MatchmakerQueueMapPoolFX bracket) {
         var min = bracket.getMinRating();
         var max = bracket.getMaxRating();
-        if (min == 0) return String.format("<%d", (int)max);
-        if (max == 0) return String.format(">%d", (int)min);
-        return String.format("%d - %d", (int)min, (int)max);
+        if (min == 0) return String.format("<%d", (int) max);
+        if (max == 0) return String.format(">%d", (int) min);
+        return String.format("%d - %d", (int) min, (int) max);
```
Comment (Brutus5000) [inline src/main/java/com/faforever/moderatorclient/ui/main_window/LadderMapPoolController.java]: Why don't you cast it to int on variable assignment? Also we don't use var in this project.
I like var in general, but would only use it consistently for new projects.

File: src/main/java/com/faforever/moderatorclient/ui/main_window/LadderMapPoolController.java [hunk 10]
```diff
@@ -176,40 +235,50 @@ private void bindListViewSelectionToSelectedMapProperty(ListView<MapVersionFX> l
...
             //bind actions for add and remove buttons
             addButton.setOnAction(event -> {
-                    mapList.add(newValue);
-                    addButton.setDisable(true);
-                    removeButton.setDisable(false);
+                MapPoolAssignmentFX newAssignment = new MapPoolAssignmentFX();
+                newAssignment.setWeight(1);
+                newAssignment.setMapParams(newValue.getMapParams());
+                newAssignment.setMapVersion(newValue.getMapVersion());
+                newAssignment.setMapPool(bracketPool);
```
Comment (Brutus5000) [inline src/main/java/com/faforever/moderatorclient/ui/main_window/LadderMapPoolController.java]: You can use the setter chaining. Each setter returns the entity itself

File: src/main/java/com/faforever/moderatorclient/ui/main_window/LadderMapPoolController.java [hunk 11]
```diff
@@ -225,4 +294,14 @@ public void onSearchMapVault() {
...
     }
 
+    public void onGeneratedMapButton() {
+        NeroxisGeneratorParams neroxisGeneratorParams = new NeroxisGeneratorParams();
+        neroxisGeneratorParams.setVersion(neroxisVersionComboBox.getValue().toString());
+        neroxisGeneratorParams.setSize(neroxisMapSizeValues[neroxisMapSizes.indexOf(neroxisSizeSpinner.getValue())]);
+        neroxisGeneratorParams.setSpawns(neroxisSpawnsSpinner.getValue());
+        MapPoolAssignment mapPoolAssignment = new MapPoolAssignment();
+        mapPoolAssignment.setMapParams(neroxisGeneratorParams);
+        selectedMap.setValue(mapPoolAssignmentMapper.map(mapPoolAssignment));
```
Comment (Brutus5000) [inline src/main/java/com/faforever/moderatorclient/ui/main_window/LadderMapPoolController.java]: You can use the setter chaining. Each setter returns the entity itself

**Review by Brutus5000 — CHANGES_REQUESTED** [review]

File: src/main/java/com/faforever/moderatorclient/ui/main_window/LadderMapPoolController.java (hunk 7 above)
Comment (Sheikah45) [reply src/main/java/com/faforever/moderatorclient/ui/main_window/LadderMapPoolController.java]: Unfortunately can't use objects.equals as Objects.equals(null,null) = true which would be true for everything as either mapParams or mapVersion is null on each assignment

File: src/main/java/com/faforever/moderatorclient/ui/main_window/LadderMapPoolController.java (hunk 8 above)
Comment (Sheikah45) [reply src/main/java/com/faforever/moderatorclient/ui/main_window/LadderMapPoolController.java]: See above

File: src/main/java/com/faforever/moderatorclient/ui/main_window/LadderMapPoolController.java (hunk 9 above)
Comment (Sheikah45) [reply src/main/java/com/faforever/moderatorclient/ui/main_window/LadderMapPoolController.java]: ah that var was already there and I didnt notice

File: src/main/java/com/faforever/moderatorclient/ui/main_window/LadderMapPoolController.java (hunk 10 above)
Comment (Sheikah45) [reply src/main/java/com/faforever/moderatorclient/ui/main_window/LadderMapPoolController.java]: Ah okay I can set it up althought it looks like it is used inconsistently on the Fx objects and I must have copied one that returned null on the setters

**Comment by Sheikah45** [comment]
seems like it could be a jdk bug https://stackoverflow.com/questions/54775253/jdk-11-0-2-compilation-fails-with-javac-npe-on-anonymous-parameterized-class-typ

**Review by Brutus5000 — APPROVED** [review]

**Pull request merged**